### ISSUE TITLE
[BUG][High priority] Lack of M84 in g-code files causes firmware crash

### ISSUE THREAD

**Issue by arekm** [prusa3d/Prusa-Firmware#3549]
**Printer type** - MK3S+
**Printer firmware version** -  3.11.0 and 3.11.1-RC1 using official hex releases

**SD card or USB/Octoprint**
Printing from SD card directly.

**Describe the bug**
Printing from g-code file (on sd card) that doesn't have M84 command causes firmware crash after processing
all existing g-code commands in file.

I've verified that all g-codes are processed by adding bunch of "M117 test" commands at the end of gcode
file to get "test "text on LCD and I got it on LCD.

**To Reproduce**
* decompress and run attached gcode file from sd card on the printer
* wait for it to finish and then observe crash 
 
**Expected behavior**
No crash.

**G-code**
This one reliably crashes for me:

[test-gcode-only-moves.gcode.gz](https://github.com/prusa3d/Prusa-Firmware/files/9198145/test-gcode-only-moves.gcode.gz)

Much less reliable but still sometimes even 3 line g-code file would also crash:
````
G90
G28 W
G1 Z50 F720 
````

**Crash dumps**
Nine crash dumps dumped by D21:
[crash-dumps.txt.gz](https://github.com/prusa3d/Prusa-Firmware/files/9198337/crash-dumps.txt.gz)

**Complete log**
[serial-crash-log.txt.gz](https://github.com/prusa3d/Prusa-Firmware/files/9200376/serial-crash-log.txt.gz)

**Video**
https://www.youtube.com/watch?v=FN9xwIY07h4

**Comment by leptun** [comment]
Just quickly looking at the crash log headers, the crashes were caused by stack overflow.
This is not a low priority issue and we've been investigating crashes for a while now. I'm tagging this for 3.12 and making it high priority.
@wavexx fyi

**Comment by leptun** [comment]
@arekm What version was used with the crash files you provided? You mentioned it happens on two official versions, but we need to know the exact release of the dump, which sadly is not included in the crash log...

**Comment by leptun** [comment]
@arekm Can you also provide a list with all the files that were on the SD card? An image of the card would be best if that is possible.

**Comment by gudnimg** [comment]
@leptun I can confirm the crash on my MK3S+. (Am using `mk3-mmu` branch though). I have some spare time atm to look at this. The crash appears when the print is done, and all gcodes are done executing.
![image](https://user-images.githubusercontent.com/8218499/181264827-549f23f5-e88d-415e-8a5f-6846c242e6c8.png)

**Comment by leptun** [comment]
I saw these messages on the stack as well in the crashes. It’s just a matter of finding where the reccursion happens in the EOF section of get_command.

**Comment by gudnimg** [comment]
I can confirm this is a recursion issue. Trying to track it down as we speak :)

Update: @leptun 
I have tracked down the recursion loop, it looks like this:
`manage_inactivity()` --> `get_command()` --> `card.printingHasFinished()` --> `st_synchronize()`--> `manage_inactivity()` (repeat)


This is the line in question: https://github.com/prusa3d/Prusa-Firmware/blob/8e72470afc38115b90058d0568b48961997235be/Firmware/cardreader.cpp#L1010 if I comment it out, the crash  is prevented.

Update 2:
The logs look like this at the end of the file (after removing the line above). I don't see any gcodes being dismissed and no recursion:
![image](https://user-images.githubusercontent.com/8218499/181282479-edb92b6a-1ed8-41ad-ad96-c8a1db2d1648.png)

**Comment by arekm** [comment]
Crash dumps are from 3.11.1-RC1 official hex build and I guess list of files/image is not needed anymore 👍 Also added [serial-crash-log.txt.gz](https://github.com/prusa3d/Prusa-Firmware/files/9200376/serial-crash-log.txt.gz)

**Comment by wavexx** [comment]
@arekm are you able to build PR #3551 and test it? Besides fixing this crash, I wonder if this has any impact on #3541

**Comment by arekm** [comment]
I'll test after weekend (don't have access to the printer now). #3541 - no way to test since no way to reproduce failing build

**Comment by arekm** [comment]
MK3 branch + mentioned fix doesn't crash on my test file but...

There seem to be another problem with it. After print finishes M27 command should return "Not SD printing" (octoprint sends M27 to monitor printing from SD) but that doesn't happen.

That's how it looks like with clean MK3 branch firmware:

````
2022-07-31 23:13:03,392 - Send: M27
2022-07-31 23:13:03,752 - Recv: T:25.3 /0.0 B:26.0 /0.0 T0:25.3 /0.0 @:0 B@:0 P:0.0 A:31.4
2022-07-31 23:13:04,145 - Recv: NORMAL MODE: Percent done: 100; print time remaining in mins: 0; Change in mins: -1
2022-07-31 23:13:04,156 - Recv: SILENT MODE: Percent done: 1; print time remaining in mins: 5; Change in mins: -1
2022-07-31 23:13:04,161 - Recv: NORMAL MODE: Percent done: 100; print time remaining in mins: 0; Change in mins: -1
2022-07-31 23:13:04,170 - Recv: SILENT MODE: Percent done: 100; print time remaining in mins: 0; Change in mins: -1
2022-07-31 23:13:05,587 - Recv: LA10C: Linear Advance mode: UNKNOWN
2022-07-31 23:13:05,591 - Recv: Done printing file
2022-07-31 23:13:05,597 - Changing monitoring state from "Printing from SD" to "Finishing"
2022-07-31 23:13:05,659 - Recv: echo:0 hours 0 minutes
2022-07-31 23:13:05,661 - Recv: LCD status changed
2022-07-31 23:13:05,709 - Recv: Not SD printing
````

and that on MK3 + 3551 PR:

````
2022-07-31 23:00:26,648 - Send: M27
2022-07-31 23:00:28,078 - Recv: NORMAL MODE: Percent done: 100; print time remaining in mins: 0; Change in mins: -1
2022-07-31 23:00:28,087 - Recv: SILENT MODE: Percent done: 1; print time remaining in mins: 5; Change in mins: -1
2022-07-31 23:00:28,096 - Recv: NORMAL MODE: Percent done: 100; print time remaining in mins: 0; Change in mins: -1
2022-07-31 23:00:28,100 - Recv: SILENT MODE: Percent done: 100; print time remaining in mins: 0; Change in mins: -1
2022-07-31 23:00:28,104 - Recv: Done printing file
2022-07-31 23:00:28,106 - Changing monitoring state from "Printing from SD" to "Finishing"
2022-07-31 23:00:28,256 - Recv: echo:0 hours 0 minutes
2022-07-31 23:00:28,258 - Recv: LCD status changed
2022-07-31 23:00:28,318 - Recv: T:26.1 /0.0 B:26.4 /0.0 T0:26.1 /0.0 @:0 B@:0 P:0.0 A:32.1
2022-07-31 23:00:29,546 - Recv: LA10C: Linear Advance mode: UNKNOWN
2022-07-31 23:00:30,323 - Recv: T:26.2 /0.0 B:26.5 /0.0 T0:26.2 /0.0 @:0 B@:0 P:0.0 A:31.7
2022-07-31 23:00:32,327 - Recv: T:26.1 /0.0 B:26.6 /0.0 T0:26.1 /0.0 @:0 B@:0 P:0.0 A:31.7
[all further messages are temperature data like earlier line]
````

No answer for M27. Octoprint then gets stuck in "Finishing" state waiting for ever for answer.

**Comment by wavexx** [comment]
@arekm I was able to reproduce this, and I believe that is probably fixed as a result of [4f22de2](https://github.com/prusa3d/Prusa-Firmware/pull/3551/commits/4f22de2333bcd9cfd603c144dd037358fc3c9d5f). Can you try again that PR?

**Comment by arekm** [comment]
Seems to be working. No crash and doesn't get stuck on M27 here.
